File: Div3-Group1/PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {

        if(head==NULL or head->next==NULL)
            return head;
        
        ListNode* lessThan=new ListNode();
        ListNode* greaterThan=new ListNode();
        ListNode* greaterThanHead=new ListNode();
        ListNode* answer= new ListNode();
        greaterThanHead->next=greaterThan;
        answer->next=lessThan;

        while(head->next!=NULL)
        {
            if(head->val<x)
            {
                lessThan->next=head;
                lessThan=lessThan->next;

            }
            else
            {
                greaterThan->next=head;
                greaterThan=greaterThan->next;

            }
            head=head->next;
        }
        if(head->val<x)
        {
            lessThan->next=head;
            lessThan=lessThan->next;
            lessThan->next=greaterThanHead->next->next;
            greaterThan->next=NULL;

        }
        else
        {
            greaterThan->next=head;
            greaterThan=greaterThan->next;
            lessThan->next=greaterThanHead->next->next;

        }

        
        return answer->next->next;
    }
};

File: leet-code-solutions/leetcode/partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        if not head:
            return None
        
        second_head = ListNode()
        answer1 = second_head

        temp = head

        while temp:
            if temp.val < x:
                second_head.next= ListNode(temp.val)
                second_head = second_head.next
            temp = temp.next

        end_of_answer1=second_head
        end_of_answer1.next=None

        second_head = ListNode()
        answer2 = second_head
  
        temp = head

        while temp:
            if temp.val >= x:
                second_head.next= ListNode(temp.val)
                second_head = second_head.next
            temp = temp.next

        second_head.next=None

        end_of_answer1.next=answer2.next
        
        return answer1.next

